### ISSUE TITLE
rdt-client Automatically Downloads Every File in a Torrent, while RD doesn't

### ISSUE THREAD

**Issue by logtho** [rogerfar/rdt-client#24]
When putting a .torrent file or magnet link in real-debrid's website, it will ask which files you want to download - with only certain files automatically checked (I assume these are the files that previous people have downloaded from that torrent - so they are the ones already cached on the server). However when downloading something through rdt-client, via sonarr or radarr at least, it automatically checks every single file in the torrent. In most cases, this results in the entire torrent having to be re-downloaded - even if it's just for an extra <1KB text file.

I am wondering if there is any way to change this so it will only download the same files as it would when putting a magnet or torrent link into RD's website, as currently I have to essentially download files to RD like you would with a torrent client, and then download those files from RD to my local machine (which in my opinion, defeats the entire purpose of RD).

Thanks.

**TL;DR:** Is there any way to make it so rdt-client will only download the same files as it would automatically download if you were to download the torrent via RD's website - rather than downloading every single file in the torrent?

**Comment by rogerfar** [comment]
I can investigate if that status is passed on through the API on the "select files" level, but in the meanwhile you can set the minimum download size in the settings dialog. I have set it to 100mb and so far I haven't had any issues yet.

**Comment by logtho** [comment]
> I can investigate if that status is passed on through the API on the "select files" level, but in the meanwhile you can set the minimum download size in the settings dialog. I have set it to 100mb and so far I haven't had any issues yet.

Thanks, this seemed like a good workaround - however I do think looking into the API to see if this is possible would be nice. Especially as this would probably leave out subtitle files (although this shouldn't be a big problem as OpenSubtitles exists! 😄)

Now I have a problem where it says "Error: The operation has timed out." when trying to download anything. I'm not quite sure what could be causing this.

It is my first time using Docker so I may have ran the command incorrectly and perhaps it is the download path (`/data/downloads`) that is incorrect? This is the command I used to install rdt-client:
```
docker run --cap-add=NET_ADMIN -d -v D:/rd-docker/:/data/downloads -v D:/rd-docker/:/data/db --log-driver json-file --log-opt max-size=10m -p 6500:6500 rogerfar/rdtclient
```
For example, I wasn't sure if I was supposed to leave those colons in the middle of the directories or not.

Sorry for the trouble.

**Comment by rogerfar** [comment]
Yes that looks OK, does it write anything in the D:/rd-docker directory? There should be a log file there and a .db file.

**Comment by logtho** [comment]
Seems that the D:/rd-docker directory is completely empty.

**Comment by rogerfar** [comment]
The premium check API is very iffy, for some reason half of the time it returns null. If you're sure you are premium then that should be fine.

But if nothing appears in that directory, does docker have the right permissions to write to it?

**Comment by logtho** [comment]
Honestly this is my first time using Docker so I'm not sure how to check that. However, I assume it should as I haven't had any issues with any other software writing to my D:\ drive. I had a quick look through Docker's settings but can't seem to find anything that would allow me to tell if it has the right permissions.

**Comment by rogerfar** [comment]
In the settings dialog there is a button to test permissions, can you check that?

**Comment by logtho** [comment]
Oh, yes, I forgot about that in the settings! It says everything looks fine:
![image](https://user-images.githubusercontent.com/78546135/106907589-504acf00-66f6-11eb-8f5f-16e812bff25a.png)

However when trying to test the download speed, it just loads for about 5-10 seconds and then stops. Not sure if this is relevant but thought I should mention it just in-case.

**Comment by rogerfar** [comment]
Yeah it's supposed to do that.

While it is doing the speed test can you check in your download folder if a file gets written and then deleted after the test is completed?

If so then everything should be correct.

**Comment by logtho** [comment]
I just ran the speed test a few times and checked in both my download path (`/data/downloads/`) and mapped path (`D:\rd-downloads`) but didn't see any files in either folders (wasn't sure which folder it was supposed to go to, so I checked both lol).

**Comment by rogerfar** [comment]
What do you mean by checked `/data/downloads/`? This is a folder inside your docker container that gets mapped to D:\rd-downloads.

**Comment by logtho** [comment]
Oh - sorry, I meant I checked the `D:/rd-docker/` directory to see if any files appeared there, as I thought that was where the `/data/downloads` directories were supposed to go. My bad.

But now that you mention it, the `D:/rd-downloads` directory is literally empty. No data folder in there. Is it supposed to be there? Should I try manually creating it myself maybe?

**Comment by rogerfar** [comment]
No it will write to the root, so you will end up with D:\rd-downloads\db.db and D:\rd-downloads\log.txt.

Can you check the logs of the container itself, restart the client then check the logs with `docker logs @rogerfar/rdtclient`

**Comment by logtho** [comment]
When I run that command I get the error: `Error: No such container: @rogerfar/rdtclient`

**Comment by logtho** [comment]
After several hours of trying to fix my above issue, I just gave up and decided to see if I can deal with downloading stuff to my C:\ drive. I thought maybe I could have it download just a few files at a time and move it to my D:\ drive with some software as soon as the download is complete.

However, even running rdt-client completely on my C:\ drive I'm still having issues. More than half of the things I download give me the same error as before; "Error: The operation has timed out." The things that do download seem to always get stuck at "Unpacking queued."
![image](https://user-images.githubusercontent.com/78546135/106950505-683a4700-6726-11eb-92ce-d6e0b34b4df4.png)

I'm wondering if you might know why this would be happening.
Sorry for asking so many questions, I'm just having a lot of trouble getting this set up 😅

**Comment by rogerfar** [comment]
The timeouts are odd, are you experiencing this too when download from RD like through a browser or download manager?

You can still use it as a download service, I might add support for that again soon as the docker is still having a lot of issues with downloading large files due to containers not liking large amounts of data passing through them.

**Comment by logtho** [comment]
> The timeouts are odd, are you experiencing this too when download from RD like through a browser or download manager?

I've always downloaded stuff through my browser and have never had any issues like that.

> You can still use it as a download service

I don't quite understand what you mean by this. Do you mean it can be used still in the same way that release 1.3 worked? If so, how do I do this as that would be perfect for me!

**Comment by logtho** [comment]
It seems like the workaround to set the minimum downloaded filesize to 100MB seems to work for most torrents with a few exceptions. I think looking into the API to see if it's possible to download only the instantly available files if possible would be a better solution in the long-term.

I've not done anything using the RD API before so I'm not entirely sure how it works, but I took a quick look at [the documentation](https://api.real-debrid.com/) and it seems like it might be possible using `/torrents/availability/{hash}`.

**Comment by logtho** [comment]
I looked into seeing if I can add this myself, but I'm not familiar with C#, your codebase, or even familiar enough with GitHub to understand how pull requests work 😅

In my opinion, this might work best if you add a toggleable setting like "only download instantly available files," which when enabled would reveal more settings. For example, in the case that there are no instantly available files, you can choose to have rdt-client either skip over the torrent and try the next one, or just have RD start downloading it (in the same way that rdt-client behaves now).

If you wanted to go even further with more customisability, it would be cool if you could set a minimum amount of seeders or a certain seeders:leechers ratio in order for rdt-client to tell RD begin the download otherwise it would skip over and try the next torrent.

I'm not sure how feasible or time-consuming this would be and I realise I'm asking a lot (especially with all the customisability settings I proposed), but in my opinion the main benefit of RD comes from having these files instantly downloadable. If you have to wait for the torrent to download to their servers before downloading it from them, it's essentially just a normal torrent client with extra steps.

I love the idea of rdt-client and appreciate the time you put into developing this, but the lack of this feature is the only thing stopping my setup from being complete. I'd love to hear your opinion on this and would love to hear about any updates of this feature. Keep up the great work!

**Comment by rogerfar** [comment]
I have implemented the /torrents/instantAvailability/ in the https://github.com/rogerfar/RD.NET.

I think you are overthinking what the RD torrent client can actually do :-) When calling the API, it only gives you a list of the files in various torrents back, no leechers or seeder info at all.

For now I will implement a feature where you can enable available files checking in conjunction with the min file size.

**Comment by Logey** [comment]
Ah, sorry. I wasn't sure exactly what rdt-client is actually capable of. But it's great news to hear that you're working on available file checking. Excited to see the results!

**Comment by logtho** [comment]
Oops, sorry. I wasn't sure what rdt-client was actually capable of. But I'm excited to hear that you're working on available file checking!

**Comment by logtho** [comment]
I see you released a new version; that was fast! I've been playing around with it and it works really well so far. I'm not sure if this issue needs to be closed now since the original purpose of the issue has been fulfilled. I'll leave it to you.

**Comment by jdotdarko** [comment]
Hi, I'm having a similar issue. Is there any way to only download a specific file to my RD cloud so that the client would only download that one file? In my case, download only the movie file and avoid extras like featurettes. 

I tried setting the minimum file size to download on the client but that doesn't work. RD cloud still downloads every file in a torrent including the files below the minimum size. What I'm trying to do is whenever I click download on Radarr, the client would send the API to RD cloud to only download the movie file which would stand out from the rest of the files inside the torrent due to its size. Is that possible?

**Comment by rogerfar** [comment]
Did you try the new version with the new setting?

**Comment by logtho** [comment]
I am using the new 1.5 release and I am still having a little trouble. The option keeps unchecking itself. If I check it, save settings, and then reopen the settings dialog, it will be unchecked.

I've tried deleting the db file and resetting rdt-client but that doesn't seem to fix it. I also tried changing the minimum file size, including setting it to 0MBs but that doesn't seem to affect it. 

I haven't messed with it enough to know if the option is actually enabled and it's just a visual bug yet though.

**Comment by jdotdarko** [comment]
> Did you try the new version with the new setting?

Yes, already tried the new version, toggling the new setting on and off.. still the same. The client downloads everything inside a torrent regardless of the minimum size set. 

In the Torrent feature of RD on its website, you can choose files what to download after entering a magnet link/torrent file. It would be great if the client auto-selects the movie file on file selection due to its obvious size and eliminate the need to for RD cloud to download because most torrents that are cached only involves the movie file and not the extras. See if everything is selected by RD cloud on file selection, most of the time it will download from torrent which would take time. This is not the case if the torrent is already cached, which instantly makes the file available for download on RDT.

**Comment by rogerfar** [comment]
@jdotdarko that's exactly what the setting "Only download available files on RealDebrid" and "Minimum file size to download" do.

If both set, it will first make a list of files that are available on Real-Debrid, then it will apply the minimum file size limit.

Version 1.5.1 has a bugfix where it didn't show the setting correctly in the settings dialog, but it should still work. I tried it on a few torrents and it skipped the files that weren't needed (with min filesize set to 0).

**Comment by jdotdarko** [comment]
> @jdotdarko that's exactly what the setting "Only download available files on RealDebrid" and "Minimum file size to download" do.
> 
> If both set, it will first make a list of files that are available on Real-Debrid, then it will apply the minimum file size limit.
> 
> Version 1.5.1 has a bugfix where it didn't show the setting correctly in the settings dialog, but it should still work. I tried it on a few torrents and it skipped the files that weren't needed (with min filesize set to 0).

It's working now! Thanks! I have one more issue tho. The client is replacing the spaces on the download link with "%20". I know this is to escape the character but is there any way for the client to at least rename it afterwards to remove the escape code?

**Comment by rogerfar** [comment]
I have added a urldecode before downloading, I will include it in the next release.
